src/vmm/src/devices/virtio/net/metrics.rs

```diff
@@ -0,0 +1,308 @@
+//!     "no_tx_avail_buffer": "SharedIncMetric",
+//!     ...
+//!  }
+//! }
+//! ```
+//! Each `net` field in the example above is a serializable `NetDeviceMetrics` structure
+//! collecting metrics such as `activate_fails`, `cfg_fails`, etc. for the network device.
+//! `net0`, `net1` and `netN` in the above example represent metrics 0th, 1st and 'N'th
+//! network device respectively and `net` is the aggregate of all the per device metrics.
+//!
+//! # Limitations
+//! Network device currently do not have `vmm::logger::metrics::StoreMetrics` so aggregate
+//! doesn't consider them.
+//!
+//! # Design
+//! The main design goals of this system are:
+//! * To improve network device metrics by logging them at per device granularity.
+//! * Continue to provide aggregate net metrics to maintain backward compatibility.
+//! * Move NetDeviceMetrics out of from logger and decouple it.
+//! * Use lockless operations, preferably ones that don't require anything other than simple
+//!   reads/writes being atomic.
+//! * Rely on `serde` to provide the actual serialization for writing the metrics.
+//! * Since all metrics start at 0, we implement the `Default` trait via derive for all of them, to
+//!   avoid having to initialize everything by hand.
+//!
+//! The system implements 1 types of metrics:
+//! * Shared Incremental Metrics (SharedIncMetrics) - dedicated for the metrics which need a counter
+//! (i.e the number of times an API request failed). These metrics are reset upon flush.
+//! We use NET_DEV_METRICS_PVT instead of adding an entry of NetDeviceMetrics
+//! in Net so that metrics are accessible to be flushed even from signal handlers.
+
+use serde::ser::SerializeMap;
+use serde::{Serialize, Serializer};
+
+use crate::logger::{IncMetric, SharedIncMetric};
+
+#[derive(Copy, Clone, Debug, Default)]
+pub struct NetDeviceMetricsIndex(usize);
+
+impl NetDeviceMetricsIndex {
+    pub fn get(&self) -> &NetDeviceMetrics {
+        // SAFETY: This is called by individual Net devices
+        // on self (i.e. self.metrics.get()) so no race condition
+        // and it is safe.
+        unsafe { &NET_DEV_METRICS_PVT.metrics[self.0] }
+    }
+}
+
+/// provides instance for net metrics
+#[derive(Debug)]
+pub struct NetDeviceMetricsPool {
+    /// used to access per net device metrics
+    metrics: Vec<NetDeviceMetrics>,
+}
+
+impl NetDeviceMetricsPool {
+    /// default construction
+    pub fn get() -> NetDeviceMetricsIndex {
+        // SAFETY: This is only called during creation of a Net device
+        // and since each device is created sequentially so there is no
+        // case of race condition.
+        unsafe {
+            NET_DEV_METRICS_PVT.metrics.push(NetDeviceMetrics::new());
+            NetDeviceMetricsIndex(NET_DEV_METRICS_PVT.metrics.len() - 1)
+        }
+    }
+}
+
+/// Contains Network-related metrics per device.
+static mut NET_DEV_METRICS_PVT: NetDeviceMetricsPool = NetDeviceMetricsPool {
+    metrics: Vec::new(),
+};
```

Comment:
I'm pretty sure this needs some sort of synchronization primitive (I would guess a `RwLock`, as writes are rare and only happen during VMM initialization, and then all the reads can essentially happen without any performance impact). Without this, concurrent reads and writes will cause undefined behavior (I havent looked at the code much, I think this can happen if Firecracker received a signal while initialization a new net device, for example. The `Vec` here will be extended without any locking, while the signal handler tries to iterate over it to dump the metrics, which will almost certainly result in a segfault). I think the safest way here is 

```
use lazy_static::lazy_static;   // we already have this dependency anyway
use std::sync::RwLock;

lazy_static! {
    /// This is an example for using doc comment attributes
    static ref METRICS: RwLock<Vec<u8>> = RwLock::new(Vec::new());
}
```
and then update all the unsafe code to instead call the correct functions on `RwLock` instead.

Although maybe I would go one step further here, because with this implementation we do not solve the unittest race condition, which is something I was really hoping we'd solve. This is just a sketch, but can we make `NET_DEV_METRICS_PVT` a `OnceCell<Vec<Arc<NetDeviceMetrics>>>` ( `NetDeviceMetrics` is `Sync` and `Send`, right? Its only atomics, so dont see why it shouldnt be). We initialize the `OnceCell` after all net devices have been created (say on `InstanceStart` or something). Then the code in the net devices will not access the global variable at all, instead each net device gets a `Arc<NetDeviceMetrics>` field (e.g. we use an `Arc` to share a reference to the `NetDeviceMetrics` object between the global and the individual net device). The signal handler and other global code that needs to dump metrics can do so by accessing the global variable, while all the devices access their own field, and do not need to ever reach out to the global state. This "fixes" our unittests, as in unittests we would simply not set up the global state at all, and only use the field that's stored in the actual device (meaning they can run in parallel).

Reply:
added the RwLock which also helped in removing all the "unsafe". Thanks for the feedback!